### PR TITLE
StringName: Use inline static field definitions

### PR DESCRIPTION
**Pull request by kisg**
Before this change StringName used regular static field definitions for its mutex, _table, configured and debug_stringname fields.

Since in the general case the ordering of the static variable and field initialization and destruction is undefined, it was possible that the destruction of StringName's static fields happened prior to the destruction of statically allocated StringName instances.

By changing the static field definitions to inline in string_name.h, the C++17 standard guarantees the correct initialization and destruction ordering.

### CODE REVIEW

**Comment by kisg** [comment]
Please consider including this in 4.3. We ran into it in a customized Godot build, but could hit anyone.

**Comment by RandomShaper** [comment]
Please give more detailed info on what's the issue this PR fixes, arising from having `StringName` instances being destroyed after `StringName::cleanup()`.

I've been checking the code and I can see `StringName::cleanup()` is called during engine termination, which sets `configured` to `false`. `StringName`'s destructor invoked later than that just early exit because of a `!configured` check. Therefore, you shouldn't even be hitting one of those `ERR_FAIL_COND(!configured)`, which may be benign but noisy.

---

Aside, regarding this changeset as a fix, as far as my understanding of `static inline` goes, the benefits boil down to not needing both a declaration and a definition. It's syntax for fusing both so definitions can happen in multiple compilation units, with the guarantee that linking will eventually keep a single copy, just like what happens with `inline` functions.

So, I'm failing to see how `static inline` can have an influence over the predictabilty of the order in which static initializers run. If anything, I can understand your compiler's implementation may be generating a different order as a side effect of duplicate folding, which results in a more favorable outcome for `StringName` by luck.

Regardless, the changes in this PR would still be good as `static inline` has already been accepted in the codebase other times, and it would make sense to fully embrace it everywhere.

**Comment by konczg** [comment]
@RandomShaper Thank you for looking at this. Yes, the current implementation of ```cleanup()``` will make sure ```configured == false```. In our custom build the engine cleanup is different, that is why the issue is not reproduced on master, we should have been more specific in our description.

As we currently see it, the current code does not introduce any actual undefined behavior, but it makes it easy to introduce one by either:
- defining a non-inline, non-local static StringName (similar to e.g. ```default_interface```) that is not default constructed,
- modifying ```StringName```'s default constructor to access any of its static variables (apart from ```configured```),
- modfying ```StringName```'s destructor so that it accesses any of its static variables before checking ```configured```.

The PR improves code robustness by fixing these potential instances of the [static initialization order fiasco](https://en.cppreference.com/w/cpp/language/siof), which has been present in Godot because of the way non-local static variables were defined by ```StringName```. Specifically, ```StringName::mutex```, ```StringName::debug_stringname``` and ```StringName::_table``` are defined as non-inline, non-local static variables, while there is a data dependency of ```StringName```'s (non-default) constructors and destructor on the initialization of these variables. This means that [ordered dynamic initialization](https://en.cppreference.com/w/cpp/language/initialization#Dynamic_initialization) of ```StringName```s must happen in the same translation unit (and in the proper sequence) where the aforementioned variables are defined, and [partially-ordered](https://en.cppreference.com/w/cpp/language/initialization#Dynamic_initialization) ```StringName```s must also be defined in the correct order in the translation units where those variables are defined; otherwise in the previously mentioned cases the construction and destruction of the ```StringName```s leads to undefined behavior.

We ran into such an issue with ```default_interface``` in a custom Godot built, where the ```StringName::StringName()``` < ```StringName::mutex``` initialization sequencing caused the destructor to lock the already deinitialized mutex, causing a crash. (In our custom build the engine cleanup is different, that is why setting configured to false is not solving the issue.)

To solve these potential issues, we can simply make the [partially-ordered dynamic initialization rule](https://en.cppreference.com/w/cpp/language/initialization#Dynamic_initialization) apply by making ```StringName```'s statics inline, therefore partially ordered: Because any definition of an (ordered or partially ordered) static ```StringName``` can only happen with a prior include of ```StringName```'s definition, and with it all the inline static definitions, this guarantees that in every translation unit the partially ordered variables are defined before all other ordered or partially ordered static definitions that depend on it, which guarantees correct sequencing.

Regarding the current use of ```inline``` in the Godot codebase: it has been used at several places already (e.g.: ```outdated_cache```, ```outdated_cache_dirty```, ```PARAM_PREFIX```, ```PARAMETERS_BASE_PATH```, ```VERSION```), and I would also advocate for the general use of it when defining non-local static variables to eliminate such SIOF issues and increase code robustness.

**Comment by RandomShaper** [comment]
Regarding the argument about how easy it is to introduce bugs by missing `configured` as a guard, well, my take is that anyone aspiring to maintain `StringName` should understand very well what they're doing and the role of each and every element involved.

Now, about the partially-ordered initialization, I can't but confess I wasn't aware of that rule. Now that I've parsed your statement as well as the references, I see why _using `inline` in these cases is even a better idea_ than I already thought.

For people following this conversation, let me state it clearly:

**We already have a bug by which `static StringName`s can use the shared `StringName` machinery before it has been set up (and/or after it has been torn down), leading to undefined behavior. If such machinery was using `inline`, the bug wouldn't exist.**

In conclusion, I don't see any downsides to this being merged and, on the contrary, I see a great deal of goodness in it. Regardless how safe and beneficial it looks, I'm not sure the issues being prevented or fixed are critical so not to defer it until 4.4 & 4.3.1.

Lastly, thank you for taking the time to explain the issue in detail!

**Review by RandomShaper — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 4.3.1.
